### PR TITLE
fix deadlock in crypto setup when it is closed while handling a message

### LINKED ISSUE #2737: Long security handshake/leaked goroutine
I noticed this goroutine on a go-ipfs node, it seems suspicious to me that after 29 minutes there's a handshake still going on but I also am not familiar with the quic-go internals.

@marten-seemann does this look like a bug, or is it potentially valid behavior?  

```
goroutine 66760407 [chan send, 29m0s]:
github.com/lucas-clemente/quic-go/internal/handshake.(*cryptoSetup).SendAlert [0xc105b86780, 0xc057f36f2a]
	pkg/mod/github.com/lucas-clemente/quic-go@v0.16.2/internal/handshake/crypto_setup.go:689
github.com/marten-seemann/qtls.(*Conn).sendAlert [0xc02220b880, 0x2a, 0x0, 0x0]
	pkg/mod/github.com/marten-seemann/qtls@v0.9.1/conn.go:838
github.com/marten-seemann/qtls.(*Conn).verifyServerCertificate [0xc02220b880, 0xc0415cedc0, 0x1, 0x1, 0x35d, 0x0]
	pkg/mod/github.com/marten-seemann/qtls@v0.9.1/handshake_client.go:877
github.com/marten-seemann/qtls.(*clientHandshakeStateTLS13).readServerCertificate [0xc015a19e20, 0x0, 0x0]
	pkg/mod/github.com/marten-seemann/qtls@v0.9.1/handshake_client_tls13.go:464
github.com/marten-seemann/qtls.(*clientHandshakeStateTLS13).handshake [0xc015a19e20, 0xc03a782700, 0x0]
	pkg/mod/github.com/marten-seemann/qtls@v0.9.1/handshake_client_tls13.go:89
github.com/marten-seemann/qtls.(*Conn).clientHandshake [0xc02220b880, 0x0, 0x0]
	pkg/mod/github.com/marten-seemann/qtls@v0.9.1/handshake_client.go:240
github.com/marten-seemann/qtls.(*Conn).Handshake [0xc02220b880, 0x0, 0x0]
	pkg/mod/github.com/marten-seemann/qtls@v0.9.1/conn.go:1395
github.com/lucas-clemente/quic-go/internal/handshake.(*cryptoSetup).RunHandshake.func1 [0xc105b86780, 0xc063a3cfc0, 0xc113900a20]
	pkg/mod/github.com/lucas-clemente/quic-go@v0.16.2/internal/handshake/crypto_setup.go:266
created by github.com/lucas-clemente/quic-go/internal/handshake. [*cryptoSetup).RunHandshak]
	pkg/mod/github.com/lucas-clemente/quic-go@v0.16.2/internal/handshake/crypto_setup.go:264
```

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #2737.

@egonelbre's analysis (option B) in https://github.com/lucas-clemente/quic-go/issues/2737#issuecomment-698514991 was right.

Unfortunately, I didn't manage to write an integration test that would trigger the bug, so I'm not 100% sure that this fix actually works under all circumstances.

### CODE REVIEW

**Comment by egonelbre** [comment]
I often try to separate the concurrency bits and channel handling into a separate struct, such that it can be separately tested... however seems bit difficult here.

If you could hook into `h.conn.Handshake` somehow, then you could call `SendAlert` from it. And concurrently call `Close`.

File: internal/handshake/crypto_setup.go [hunk 1]
```diff
@@ -506,11 +507,12 @@ func (h *cryptoSetup) ReadHandshakeMessage() ([]byte, error) {
 	} else {
```
Comment (egonelbre) [inline internal/handshake/crypto_setup.go:507]: On line 488, is there a possibility the `go func() { <-h.isReadingHandshakeMessage` gets stuck somehow?

**Comment by marten-seemann** [comment]
@egonelbre Yes, that would be nice, but the way the TLS library is structured doesn't allow that.
Ideally, the library would expose a `HandleHandshakeMessage() (response []byte, err error)` method. Instead, we need to call `Handshake`, which basically does the following:
```go
func Handshake() {
    for {
          msg := cryptosetup.ReadHandshakeMessage()
          if err := handleMsg(msg); err != nil {
               cryptosetup.SendAlert()
               break
          }
          if handshake complete {
               break
          }
    }
}

File: internal/handshake/crypto_setup.go (hunk 1 above)
Comment (marten-seemann) [reply internal/handshake/crypto_setup.go:507]: Good catch. That one should probably be protected by a `select` statement as well.

**Review by lucas-clemente — APPROVED** [review]

**Pull request merged**